include/beast/static_string.hpp

```diff
@@ -0,0 +1,696 @@
+{
+    if(pos >= n_)
+        throw std::out_of_range("static_string::at");
+    return s_[pos];
+}
+
+template<std::size_t N, class CharT, class Traits>
+auto
+static_string<N, CharT, Traits>::
+at(size_type pos) const ->
+    const_reference
+{
+    if(pos >= n_)
+        throw std::out_of_range("static_string::at");
+    return s_[pos];
+}
+
+template<std::size_t N, class CharT, class Traits>
+void
+static_string<N, CharT, Traits>::
+resize(std::size_t n)
+{
+    if(n > N)
+        throw std::length_error("static_string overflow");
+    n_ = n;
+    s_[n_] = 0;
+}
+
+template<std::size_t N, class CharT, class Traits>
+void
+static_string<N, CharT, Traits>::
+resize(std::size_t n, CharT c)
+{
+    if(n > N)
+        throw std::length_error("static_string overflow");
+    if(n > n_)
+        Traits::assign(&s_[n_], n - n_, c);
+    n_ = n;
+    s_[n_] = 0;
+}
+
+template<std::size_t N, class CharT, class Traits>
+template<std::size_t M>
+int
+static_string<N, CharT, Traits>::
+compare(static_string<M, CharT, Traits> const& rhs) const
+{
+    if(size() < rhs.size())
+    {
+        auto const v = Traits::compare(
+            data(), rhs.data(), size());
+        if(v == 0)
+            return -1;
+        return v;
+    }
+    else if(size() > rhs.size())
+    {
+        auto const v = Traits::compare(
+            data(), rhs.data(), rhs.size());
+        if(v == 0)
+            return 1;
+        return v;
+    }
+    return Traits::compare(data(), rhs.data(), size());
+}
+
+template<std::size_t N, class CharT, class Traits>
+void
+static_string<N, CharT, Traits>::
+assign(CharT const* s)
+{
+    auto const n = Traits::length(s);
+    if(n > N)
+        throw std::out_of_range("too large");
+    n_ = n;
+    Traits::copy(&s_[0], s, n_ + 1);
+}
+
+namespace detail {
+
+template<std::size_t N, std::size_t M, class CharT, class Traits>
+int compare(
+    static_string<N, CharT, Traits> const& lhs,
+    const CharT (&s)[M])
+{
+    if(lhs.size() < M-1)
+    {
+        auto const v = Traits::compare(
+            lhs.data(), &s[0], lhs.size());
+        if(v == 0)
+            return -1;
+        return v;
+    }
+    else if(lhs.size() > M-1)
+    {
+        auto const v = Traits::compare(
+            lhs.data(), &s[0], M-1);
+        if(v == 0)
+            return 1;
+        return v;
+    }
+    return Traits::compare(lhs.data(), &s[0], lhs.size());
+}
+
+template<std::size_t N, std::size_t M, class CharT, class Traits>
+int compare(
+    const CharT (&s)[M],
+    static_string<N, CharT, Traits> const& rhs)
+{
+    if(M-1 < rhs.size())
+    {
+        auto const v = Traits::compare(
+            &s[0], rhs.data(), M-1);
+        if(v == 0)
+            return -1;
+        return v;
+    }
+    else if(M-1 > rhs.size())
+    {
+        auto const v = Traits::compare(
+            &s[0], rhs.data(), rhs.size());
+        if(v == 0)
+            return 1;
+        return v;
+    }
+    return Traits::compare(&s[0], rhs.data(), M-1);
```

Comment:
Silly question: why not implement this as `return -compare(rhs, s);` to call the other override, and reduce code duplication (particularly if you declare this override as `inline`)?